x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/AnalyticsCollectionService.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.entsearch.analytics;
+
+import org.elasticsearch.ResourceAlreadyExistsException;
+import org.elasticsearch.ResourceNotFoundException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.datastreams.CreateDataStreamAction;
+import org.elasticsearch.action.datastreams.DeleteDataStreamAction;
+import org.elasticsearch.action.support.IndicesOptions;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.client.internal.OriginSettingClient;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.service.ClusterService;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ENT_SEARCH_ORIGIN;
+
+/**
+ * A service that allows the manipulation of persistent {@link AnalyticsCollection} model.
+ *
+ * Until we have more specific need the {@link AnalyticsCollection} is just another representation
+ * of a {@link org.elasticsearch.cluster.metadata.IndexAbstraction.DataStream}.
+ *
+ * As a consequence, this service is mostly a facade for the data stream API.
+ */
+public class AnalyticsCollectionService implements ClusterStateListener {
+
+    private final Client clientWithOrigin;
+
+    private final IndexNameExpressionResolver indexNameExpressionResolver;
+
+    private Map<String, AnalyticsCollection> analyticsCollections = Collections.emptyMap();
+
+    public AnalyticsCollectionService(
+        Client client,
+        ClusterService clusterService,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        this.clientWithOrigin = new OriginSettingClient(client, ENT_SEARCH_ORIGIN);
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+        clusterService.addListener(this);
+    }
+
+    /**
+     * Retrieve an analytics collection by name {@link AnalyticsCollection}
+     *
+     * @param collectionName {@link AnalyticsCollection} name.
+     * @param listener The action listener to invoke on response/failure.
+     */
+    public void getAnalyticsCollection(String collectionName, ActionListener<AnalyticsCollection> listener) {
+        if (analyticsCollections.containsKey(collectionName) == false) {
```

Comment:
We shouldn't access this map directly since it can be updated by the cluster change listener. Let's use a temp variable first to ensure that we're doing all the lookups on the same object?

Reply:
Now using a temp variable.